### PR TITLE
feat(math): introduce math module with Point class for 2D coordinates

### PR DESCRIPTION
**Pull request by dkrutskikh**
### **PR Type**
Enhancement, Tests

___

### **Description**
- Introduce `Point` class for 2D integer coordinate representation

- Add comprehensive `Point` methods including arithmetic operators and utility functions

- Implement full test suite with 560 lines covering constructors, operators, and edge cases

- Add `[[nodiscard]]` attributes to `FixedString` accessor methods for better code safety

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["mathematics.hpp<br/>Module Header"] --> B["Point Class<br/>2D Integer Coordinates"]
  B --> C["point.hpp<br/>Class Declaration"]
  B --> D["point.inl<br/>Method Implementations"]
  E["FixedString<br/>Accessor Methods"] --> F["[[nodiscard]]<br/>Attributes"]
  C --> G["Test Suite<br/>560 Lines"]
  D --> G
  H["CMakeLists.txt<br/>Build Configuration"] --> I["Mathematics<br/>Source Group"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>mathematics.hpp</strong><dd><code>Create mathematics module header</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

include/mathematics.hpp

<ul><li>New header file establishing mathematics namespace and module <br>structure<br> <li> Defines <code>real_t</code> type alias as <code>float</code> for floating-point operations<br> <li> Includes <code>Point</code> class header and inline implementation files</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/159/files#diff-04455ec0b235aa2037c4fd0abf5937acd7847a58653476f50e6b42fa035e6d99">+40/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>point.hpp</strong><dd><code>Define Point class interface and API</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

include/mathematics/point.hpp

<ul><li>Defines <code>Point</code> class with <code>std::int32_t</code> value type for 2D coordinates<br> <li> Declares three constructors: default, coordinate-based, and <br>array-based<br> <li> Implements arithmetic operators: <code>+=</code>, <code>-=</code>, <code>*=</code> (integer and real <br>scalars), <code>/=</code> (integer and real scalars)<br> <li> Provides utility methods: <code>c_arr()</code>, <code>sqrMagnitude()</code>, <code>setZero()</code>, <br><code>isZero()</code>, <code>isEqual()</code> with tolerance support<br> <li> All methods marked as <code>constexpr</code> and <code>[[nodiscard]]</code> where appropriate</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/159/files#diff-80cbffe132c6b3a68f56c70f618532b0837385a354992c719a8ddfb677e151e1">+241/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>point.inl</strong><dd><code>Implement Point class methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

include/mathematics/point.inl

<ul><li>Implements all <code>Point</code> class methods as <code>constexpr</code> for compile-time <br>evaluation<br> <li> Provides array access via <code>c_arr()</code> returning pointer to x coordinate<br> <li> Implements arithmetic operations with both integer and floating-point <br>scalars<br> <li> Includes validation assertions for null pointers and non-zero divisors<br> <li> Implements comparison and utility methods with tolerance support</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/159/files#diff-c18606bb9a02b4d2d32b5b5ff55a4bbf07de4fe7f3da4e4e9b6aa3025bae9a17">+118/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>fixed_string.hpp</strong><dd><code>Add nodiscard attributes to accessors</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

include/core/fixed_string.hpp

<ul><li>Add <code>[[nodiscard]]</code> attribute to <code>front()</code> method to encourage result <br>usage<br> <li> Add <code>[[nodiscard]]</code> attribute to <code>back()</code> method to encourage result usage<br> <li> Add <code>[[nodiscard]]</code> attribute to <code>data()</code> method to encourage result usage<br> <li> Improves code safety by warning when return values are discarded</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/159/files#diff-13f6b77c1e124b2465fc0ef8b977ff9646f80c1c741058c0f118d62d17278bc3">+3/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>point.cpp</strong><dd><code>Add comprehensive Point class tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

test/mathematics/point.cpp

<ul><li>Comprehensive test suite with 560 lines covering all <code>Point</code> <br>functionality<br> <li> Tests constructors with various coordinate combinations and edge cases<br> <li> Validates <code>c_arr()</code> method for both const and non-const access<br> <li> Tests all arithmetic operators with chained operations<br> <li> Verifies utility methods including <code>sqrMagnitude()</code>, <code>setZero()</code>, <br><code>isZero()</code>, and <code>isEqual()</code> with tolerance<br> <li> Includes both compile-time (<code>STATIC_REQUIRE</code>) and runtime tests</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/159/files#diff-0fdf344105e0bf5b5025da5cc34cfbaaff04e0641e276c9a9f872b7b3242fde5">+560/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>CMakeLists.txt</strong><dd><code>Integrate mathematics module into build system</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

CMakeLists.txt

<ul><li>Create new <code>SRC_MATHEMATICS_LIST</code>, <code>HDR_MATHEMATICS_LIST</code>, and <br><code>INL_MATHEMATICS_LIST</code> variables<br> <li> Add mathematics source group organization in project structure<br> <li> Integrate mathematics module files into main library build <br>configuration<br> <li> Maintain consistent formatting with existing Core module structure</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/159/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20a">+9/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: include/core/fixed_string.hpp [hunk 1]
```diff
@@ -430,7 +430,7 @@ class FixedString {
     \note The returned reference allows modification of the character.
     \note Use const version for read-only access.
   */
-  constexpr char & front() noexcept;
+  [[nodiscard]] constexpr char & front() noexcept;
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply include/core/fixed_string.hpp]: fixed

File: include/mathematics/point.hpp [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    This method calculates the squared magnitude (length) of the vector represented by this point.
+    The squared magnitude is calculated as x² + y².
+
+    \return The squared magnitude of this point.
+
+    \note This method avoids the expensive square root operation by returning the squared value.
+    \note Use this method when comparing magnitudes for performance reasons.
+  */
+  [[nodiscard]] constexpr value_type sqrMagnitude() const noexcept;
+
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply include/mathematics/point.hpp]: won't be fixed

File: include/mathematics/point.inl [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+constexpr Point & Point::operator/=(real_t scalar) noexcept {
+  assert_message(scalar > 0 || scalar < 0, "scalar must be non-zero");
+
+  x = static_cast<value_type>(x / scalar);
+  y = static_cast<value_type>(y / scalar);
+
+  return *this;
+}
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply include/mathematics/point.inl]: won't be fixed

File: include/mathematics/point.inl [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+constexpr bool Point::isZero() const noexcept {
+  return x == 0 && y == 0;
+}
+
+constexpr bool Point::isEqual(const Point & point, value_type tolerance) const noexcept {
+  assert_message(tolerance >= 0, "tolerance must be non-negative");
+
+  return (abs(x - point.x) <= tolerance) && (abs(y - point.y) <= tolerance);
+}
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply include/mathematics/point.inl]: fixed

File: test/mathematics/point.cpp [hunk 5]
```diff
@@ -0,0 +1,560 @@
...
+  SECTION("Constructor from array") {
+    constexpr Point point({42, -17});
+
+    REQUIRE(point.x == 42);
+    REQUIRE(point.y == -17);
+
+    // Compile-time checks
+    STATIC_REQUIRE(point.x == 42);
+    STATIC_REQUIRE(point.y == -17);
+  }
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply test/mathematics/point.cpp]: fixed

File: include/math/point.inl [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+constexpr bool Point::isZero() const noexcept {
+  return x == 0 && y == 0;
+}
+
+constexpr bool Point::isEqual(const Point & point, value_type tolerance) const noexcept {
+  assert_message(tolerance >= 0, "tolerance must be non-negative");
+
+  return (std::abs(x - point.x) <= tolerance) && (std::abs(y - point.y) <= tolerance);
+}
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply include/math/point.inl]: won't be fixed

File: include/math.hpp [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+#ifndef INCLUDE_MATH_HPP_
+#define INCLUDE_MATH_HPP_
+
+/// @namespace toy::math
+/// @brief Contains all public mathematical types and utilities of ToyGine.
+namespace toy::math {
+
+using real_t = float; //!< floating point number type.
+
+} // namespace toy::math
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply include/math.hpp]: fixed

File: include/math/point.hpp [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+
+    \pre The \a tolerance must be non-negative.
+
+    \note When tolerance is 0, this performs exact equality comparison.
+    \note When tolerance is greater than 0, this performs approximate equality comparison.
+  */
+  [[nodiscard]] bool isEqual(const Point & point, value_type tolerance = 0) const noexcept;
+};
+
+} // namespace toy::math
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply include/math/point.hpp]: fixed

File: include/math/point.inl [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+// PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+// FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+//
+/*!
+  \file   point.inl
+  \brief  TODO
+*/
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply include/math/point.inl]: fixed

File: include/math/point.inl [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+  y = values[1];
+}
+
+constexpr Point::pointer Point::c_arr() noexcept {
+  return &x;
+}
+
+constexpr Point::const_pointer Point::c_arr() const noexcept {
+  return &x;
+}
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply include/math/point.inl]: fixed

File: test/math/point.cpp [hunk 11]
```diff
@@ -0,0 +1,566 @@
...
+    // Test coordinate constructor at runtime
+    Point coordPoint(123, -456);
+    REQUIRE(coordPoint.x == 123);
+    REQUIRE(coordPoint.y == -456);
+
+    // Test array constructor at runtime
+    Point arrayPoint({789, -321});
+    REQUIRE(arrayPoint.x == 789);
+    REQUIRE(arrayPoint.y == -321);
+  }
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply test/math/point.cpp]: fixed
